File: src/albumexplore/gui/utils/tag_filter.py
"""
Simple TagFilter implemented in Python using BitSet for demo/prototyping.

API:
  - TagFilter(tag_map: Dict[str, BitSet])
  - evaluate(expr) -> BitSet  (expr is a small AST built as tuples/dicts)
  - matching_counts_for_visible(expr, visible_tags) -> list of {tag, count}
  - samples_for(expr, limit)

This mirrors the TypeScript TagFilter we added earlier so UI code can be
implemented either in Python or TS depending on the view.
"""
from typing import Dict, List, Optional
from .bitset import BitSet


TagMap = Dict[str, BitSet]


class TagFilter:
    def __init__(self, tag_map: TagMap):
        self.tag_map = tag_map

    def evaluate(self, expr) -> BitSet:
        if expr is None:
            # empty expression -> universe
            return self.universe()
        t = expr.get('type')
        if t == 'tag':
            return self.tag_map.get(expr.get('tag'), BitSet())
        elif t == 'and':
            return self.evaluate(expr['left']).intersect(self.evaluate(expr['right']))
        elif t == 'or':
            return self.evaluate(expr['left']).union(self.evaluate(expr['right']))
        elif t == 'not':
            return self.universe().difference(self.evaluate(expr['expr']))
        else:
            return BitSet()

    def universe(self) -> BitSet:
        out: Optional[BitSet] = None
        for k in self.tag_map:
            if out is None:
                out = self.tag_map[k].clone()
            else:
                out = out.union(self.tag_map[k])
        return out or BitSet()

    def matching_counts_for_visible(self, expr, visible_tags: List[str]):
        base = self.evaluate(expr) if expr is not None else None
        out = []
        for tag in visible_tags:
            b = self.tag_map.get(tag, BitSet())
            c = base.intersect(b).size() if base is not None else b.size()
            out.append({'tag': tag, 'count': c})
        return out

    def samples_for(self, expr, limit=5) -> List[int]:
        s = self.evaluate(expr) if expr is not None else self.universe()
        return s.to_list()[:limit]

File: src/albumexplore/gui/utils/TagFilter.ts
// src/albumexplore/gui/utils/TagFilter.ts
// Small utility for evaluating boolean tag expressions using BitSet

import BitSet from './BitSet';

export type TagMap = { [tag: string]: BitSet };

// Expression AST types
export type Expr =
    | { type: 'tag'; tag: string }
    | { type: 'and'; left: Expr; right: Expr }
    | { type: 'or'; left: Expr; right: Expr }
    | { type: 'not'; expr: Expr };

export default class TagFilter {
    private tagMap: TagMap;

    constructor(tagMap: TagMap) {
        this.tagMap = tagMap;
    }

    evaluate(expr: Expr): BitSet {
        switch (expr.type) {
            case 'tag':
                return this.tagMap[expr.tag] ? this.tagMap[expr.tag].clone() : new BitSet(0);
            case 'and': {
                const L = this.evaluate(expr.left);
                const R = this.evaluate(expr.right);
                return L.intersect(R);
            }
            case 'or': {
                const L = this.evaluate(expr.left);
                const R = this.evaluate(expr.right);
                return L.union(R);
            }
            case 'not': {
                // not is taken relative to the universe: compute universe bitset as union of all tags
                const universe = this.universe();
                const inner = this.evaluate(expr.expr);
                return universe.difference(inner);
            }
            default:
                return new BitSet(0);
        }
    }

    // quickly compute universe (cacheable optimization could be added)
    universe(): BitSet {
        const tags = Object.keys(this.tagMap);
        if (tags.length === 0) return new BitSet(0);
        let out = this.tagMap[tags[0]].clone();
        for (let i = 1; i < tags.length; i++) out = out.union(this.tagMap[tags[i]]);
        return out;
    }

    // helper: compute Matching counts for a set of candidate tags, returning { tag, count }
    matchingCountsForVisible(expr: Expr | null, visibleTags: string[]): { tag: string; count: number }[] {
        const base = expr ? this.evaluate(expr) : null;
        const out: { tag: string; count: number }[] = [];
        for (const tag of visibleTags) {
            const b = this.tagMap[tag] || new BitSet(0);
            const c = base ? base.intersect(b).size() : b.size();
            out.push({ tag, count: c });
        }
        return out;
    }

    // get samples (album ids) for an expression (limit)
    samplesFor(expr: Expr | null, limit = 5): number[] {
        const set = expr ? this.evaluate(expr) : this.universe();
        return set.toArray().slice(0, limit);
    }
}
